### PR TITLE
DBZ-2790 Implement default values for postgres

### PR DESCRIPTION
**Pull request by kppullin**
https://issues.redhat.com/browse/DBZ-2790

Using the `SqlServerDefaultValueConverter` class as a starting point, this commit implements default value support for postgres.

### CODE REVIEW

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresConnection.java [hunk 1]
```diff
@@ -62,18 +63,25 @@
...
      *
      * @param config {@link Configuration} instance, may not be null.
-     * @param provideTypeRegistry {@code true} if type registry should be created
+     * @param typeRegistry {@link TypeRegistry} the type registry to bind to this connection.
+     * @param valueConverter {@link PostgresValueConverter} a postgres value converter.
      */
-    public PostgresConnection(Configuration config, boolean provideTypeRegistry) {
+    public PostgresConnection(Configuration config, TypeRegistry typeRegistry, PostgresValueConverter valueConverter) {
         super(config, FACTORY, PostgresConnection::validateServerVersion, PostgresConnection::defaultSettings);
-        this.typeRegistry = provideTypeRegistry ? new TypeRegistry(this) : null;
```
Comment (kppullin) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresConnection.java:76]: I had a heck of a time trying to tease apart the dependencies between `PostgresConnection`, `TypeRegistry`, and `PostgresValueConverter`.  Open to alternative suggestions!

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java [hunk 2]
```diff
@@ -370,6 +374,91 @@ public void shouldPopulateToastableColumnsCache() throws Exception {
...
+                // money
+                "numeric NUMERIC(10, 5) default 12345.67891, " +
+                // path
+                // pg_lsn
+                // point
+                // polygon
+                "real FLOAT4 default 1234567890.5, " +
+                "smallint INT2 default 32767, " +
+                "text TEXT default 'asdf', " +
+                // time
```
Comment (kppullin) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java:411]: Most notably the date & time related types are omitted here.  At least in my case they're all expressions using either `now()` or `current_timestamp`, which I don't _think_ will translate well as default values.  If desired I can add code to at least attempt to parse the provided default in case it is a valid date and let it fail otherwise.

**Comment by gunnarmorling** [comment]
Hey @kppullin, thanks a lot for this contribution! I couldn't look into this PR yet; can you perhaps add a high-level description of the design of the change? In particular, is it prone to inconsistencies if the default value changes multiple times in a short period of time?

**Comment by kppullin** [comment]
In terms of high level design, this is more or less a port of the Sql Server default handling code, with changes so that the postgres specific types & default formats are handled, and I lack much deeper insights into the internal mechanics beyond this.

I'm actually not quite sure how'd it react upon rapid multiple changes... looking at how changes to the schema are handled in the ChangeRecordEmitters it _seems_ that it wouldn't be an issue however this is completely not validated by any manual or code tests.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/TypeRegistry.java [hunk 3]
```diff
@@ -98,7 +98,7 @@
     private final Map<String, PostgresType> nameToType = new HashMap<>();
     private final Map<Integer, PostgresType> oidToType = new HashMap<>();
 
-    private final PostgresConnection connection;
+    private PostgresConnection connection;
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/TypeRegistry.java:101]: I'm feeling a bit uneasy about this one; why is this needed?

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java:411]: Yes, if an actual value is provided, it'd be good to propagate this (I believe we do this for SQL Server, too). `now()` and related non-absolute values are problematic of course, I believe we export the equivalent to zero in this case for SQL Server. @jpechane, can you confirm?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresDefaultValueConverter.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            if ((result instanceof BigDecimal) && column.scale().isPresent() && column.scale().get() > ((BigDecimal) result).scale()) {
+                // Note that as the scale is increased only, the rounding is more cosmetic.
+                result = ((BigDecimal) result).setScale(column.scale().get(), RoundingMode.HALF_EVEN);
+            }
+            return result;
+        }
+        return defaultValue;
+    }
+
+    private Map<String, DefaultValueMapper> createDefaultValueMappers() {
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresDefaultValueConverter.java:110]:
```suggestion
    private static Map<String, DefaultValueMapper> createDefaultValueMappers() {
```

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresDefaultValueConverter.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+        result.put("json", v -> extractDefault(v)); // Sample value: '{}'::json
+        result.put("jsonb", v -> extractDefault(v)); // Sample value: '{}'::jsonb
+        result.put("xml", v -> extractDefault(v)); // Sample value: '<foo>bar</foo>'::xml
+
+        // Other data types, such as box, bytea, date, time and more are not handled.
+        return result;
+    }
+
+    private String extractDefault(String defaultValue) {
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresDefaultValueConverter.java:144]:
```suggestion
    private static String extractDefault(String defaultValue) {
```

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresDefaultValueConverter.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        // Values are either "raw", such as `1234`, or "type casted", such as `'9223372036854775807'::bigint`.
+        // If the value does NOT contain a single quote it is assumed to be a raw value. Otherwise the value is
+        // extracted from inside the single quotes.
+        if (!defaultValue.contains("'")) {
+            return defaultValue;
+        }
+
+        final Matcher matcher = Pattern.compile("'(.*?)'").matcher(defaultValue);
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresDefaultValueConverter.java:152]: Let's extract the Pattern into a constant and re-use it.

**Review by gunnarmorling — COMMENTED** [review]
Hey, sorry for the super-long delay :(

Overall, this LGTM, a few questions on some details inline. On a general note, I suppose this will be prone to potential inconsistencies, when it comes to _changes_ of a column's default value. E.g. a sequence of events like this:

1. Change Event 1 for column "foo" with default value "A"
2. Modify "foo"'s default value from "A" to "B"
3. Change Event 2 for column "foo"

I suppose the connector would not be aware of that schema change and export both events, 1 and 2, with the same default value of "A"?

The same if a default value gets dropped or added while the connector is running? Also, after a restart, the connector would pick up any changed default values, which also may lead to exporting "old" events from before that default value switch with the updated default value.

If that's the case indeed, I think it's an acceptable limitation (and not one we can bypass, in all likelyhood), it'd deserve mentioning in the docs though.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/TypeRegistry.java (hunk 3 above)
Comment (kppullin) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/TypeRegistry.java:101]: Agreed, I wasn't really happy with how I hacked this part together.  It's been a while and I'm trying to recall the specifics, but I somewhat referenced the issue with this comment: https://github.com/debezium/debezium/pull/1973/files#r529883357

This change was "forced" by trying to follow the existing SqlServer pattern where the `SqlServerValueConverter` is injected into the `SqlServerConnection`'s constructor.  However, the postgres code has an additional bit of complexity via the `TypeRegistry`.

The circular dependency between the `PostgresConnection` and `TypeRegistry` caused friction along with `PostgresValueConverter`, in which I worked around by first creating the `TypeRegistry` independently of the connection so that the registry can be injected into the converter, and then injected the registry into the connection (in which the connection then calls `prime()` to do what the registry's constructor did previously).

Basically the mess I created here: https://github.com/debezium/debezium/pull/1973/files#diff-0c1ba77f25c31166ad8d358987cf467ba70689a3cd15b58f0d486959fdc99dddR77-R94 and https://github.com/debezium/debezium/pull/1973/files#diff-f2d4d5bd7e1a3d93fe13e699b7ec7ba0fb138468c6c35b1592e08449142e87a7R82-R84

Without sufficient understanding of the rest of the code I tried to keep my changes minimal as to not break anything else by attempting a larger refactoring.  Sadly, I'm lacking the time now to drive this further even if given some additional pointers on a better design.

**Comment by kppullin** [comment]
> Hey, sorry for the super-long delay :(

No worries!  We ended up not using this change since many of our columns depend on `now()` style defaults, which don't work with this model too well.  Instead we've accepted defeat and enabled the `NONE` schema registry compatibility setting (the original motivation for this change was to try to keep some level of compatibility during schema changes), working are the limitation elsewhere in our stack.

Given the above and a lack of time to pursue this further, you may wish to close out this PR and leave it as reference in case someone else picks up the effort.

**Comment by eslep** [comment]
> ...leave it as reference in case someone else picks up the effort

Maybe that time has come 😄 

Hey y'all, so I find myself in a similar situation to @kppullin it seems... needing either `DEFAULT` support or to configure `NONE` compatibility on the schemas from Debezium. The approach here is the same as I had in mind when I checked the code, so thanks a lot for leaving this here :) I'd love to try to pick up the torch and finish what is remaining here, so I wanted to check in and see if I understood the remaining to-do list:
* Tests to show behaviour when rapid changes are made to column default; docs for the observed behaviour
* Propagate absolute-valued defaults for 'time' columns, or a 'throwaway' value for any non-absolute column defaults
* Extract Pattern in `PostgresDefaultValueConverter` to a constant
* See if there is a way to address https://github.com/debezium/debezium/pull/1973#discussion_r570149115

This is what I understood from the comments, have I missed anything?
I would proceed based on this for now, once I have the chance to start, and just adjust if you have corrections to make to that list.
When the time comes, should I open a new PR and reference this one as the original? (I don't assume I can continue this one directly... or am I wrong?)

**Comment by gunnarmorling** [comment]
Hey @eslep, sorry for the radio silence. Your contribution will be definitely welcomed for wrapping up this one, should you still be interested. Your description of open tasks sounds like a great starting point.

> When the time comes, should I open a new PR and reference this one as the original?

Yes, please open a new PR. This definitely also needs rebasing by now. Thanks a lot!

**Comment by eslep** [comment]
@gunnarmorling no worries! I've anyways not been able to wrap this up just yet, so the delay was no issue. I have one test case left to finish, plus the docs... so hopefully you will see a new PR in the near future 😉

**Comment by gunnarmorling** [comment]
Hey @eslep, any news here?

**Comment by eslep** [comment]
@gunnarmorling Yes, sorry for the delay! With the last test case I was working on, I uncovered a small bug and needed to fix that, along with integrating your changes from upstream. I am creating the PR right now while I work on the docs, just wanted to get some eyes on the implementation before I wrap it up completely

**Comment by kppullin** [comment]
Closing in favor of #2339
